### PR TITLE
feat(artifact): add OCI references support

### PR DESCRIPTION
**Pull request by rchincha**
Signed-off-by: Ramkumar Chinchani <rchincha@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
```
# artifact media-type
regctl artifact put --media-type "application/vnd.oci.artifact.manifest.v1+json" --annotation demo=true --annotation format=oci --artifact-type "application/vnd.example.icecream.v1" localhost:8080/artifact:demo <<EOF
test artifact
EOF

regctl manifest get localhost:8080/artifact:demo

regctl artifact get localhost:8080/artifact:demo

regctl artifact list localhost:8080/artifact:demo

regctl artifact put --subject localhost:8080/artifact:demo << EOF
test reference
EOF

regctl artifact list localhost:8080/artifact:demo
```

**Comment by rchincha** [comment]
```
+ regctl artifact list --filter-artifact-type application/vnd.example.icecream.v1 localhost:8080/artifact:demo
Subject:                                        localhost:8080/artifact:demo

Referrers:

  Name:                                         localhost:8080/artifact@sha256:d1516843be4ee55eefd1a2a3edcb4eb98e49e638e8f914490a5bcd70231b2aff
  Digest:                                       sha256:d1516843be4ee55eefd1a2a3edcb4eb98e49e638e8f914490a5bcd70231b2aff
  MediaType:                                    application/vnd.oci.image.manifest.v1+json
  ArtifactType:                                 application/vnd.example.icecream.v1
  Annotations:
    demo:                                       true
    format:                                     oci
Annotations:
  org.opencontainers.references.filtersApplied:
  ```

**Comment by sudo-bmitch** [comment]
> + regctl artifact list localhost:8080/artifact:demo
> Subject:        localhost:8080/artifact:demo
> 
> Referrers:
> 
>   Name:         localhost:8080/artifact@sha256:d1516843be4ee55eefd1a2a3edcb4eb98e49e638e8f914490a5bcd70231b2aff
>   Digest:       sha256:d1516843be4ee55eefd1a2a3edcb4eb98e49e638e8f914490a5bcd70231b2aff
>   MediaType:    application/vnd.oci.image.manifest.v1+json
>   ArtifactType: application/vnd.example.icecream.v1
>   Annotations:
>     format:     oci
>     demo:       true

That output looks reversed. The listing of artifacts that refer to demo shouldn't be the demo artifact itself. From my own testing:

```
$ regctl image copy localhost:5000/regclient/regctl:edge localhost:5207/regclient/regctl:edge                                                                                                                      

$ echo "hello artifact" | regctl artifact put --artifact-type application/example.demo --media-type application/vnd.oci.artifact.manifest.v1+json localhost:5207/regclient/regctl:artifact

$ echo "hello image" | regctl artifact put --artifact-type application/example.demo localhost:5207/regclient/regctl:image                                                                                                                                                                                                   

$ echo "hello subject" | regctl artifact put --artifact-type application/example.demo --subject localhost:5207/regclient/regctl:edge                                                        

$ regctl tag ls localhost:5207/regclient/regctl                 
artifact                                                                            
edge                                                                               
image                                                                             
                                                                                       
$ regctl artifact list localhost:5207/regclient/regctl:edge
Subject:   localhost:5207/regclient/regctl:edge              
                                                                                       
Referrers:                                                                       

$ regctl artifact get localhost:5207/regclient/regctl:image
hello image

$ regctl artifact get localhost:5207/regclient/regctl:artifact
hello artifact
```

The listing of artifacts that point to edge shouldn't be empty.

**Comment by sudo-bmitch** [comment]
Latest change works for me. 👍🏼

File: pkg/storage/local/local.go [hunk 1]
```diff
@@ -1339,7 +1340,120 @@ func (is *ImageStoreLocal) DeleteBlob(repo string, digest godigest.Digest) error
...
+	index = ispec.Index{
+		Versioned:   imeta.Versioned{SchemaVersion: defaultSchemaVersion},
+		MediaType:   ispec.MediaTypeImageIndex,
+		Manifests:   result,
+		Annotations: map[string]string{},
+	}
+
+	// response was filtered by artifactType
+	if artifactType != "" {
+		index.Annotations[storageConstants.ReferrerFilterAnnotation] = ""
```
Comment (peusebiu) [inline pkg/storage/local/local.go:1449]: shouldn't this be index.Annotations[storageConstants.ReferrerFilterAnnotation] = artifactType ?

File: pkg/storage/s3/s3.go [hunk 2]
```diff
@@ -1224,7 +1224,12 @@ func (is *ObjectStorage) GetBlobContent(repo string, digest godigest.Digest) ([]
 	return buf.Bytes(), nil
 }
 
-func (is *ObjectStorage) GetReferrers(repo string, digest godigest.Digest, mediaType string,
+func (is *ObjectStorage) GetReferrers(repo string, digest godigest.Digest, artifactType string,
```
Comment (andaaron) [inline pkg/storage/s3/s3.go:1227]: We should add a task to track implementing this?

File: pkg/api/routes.go [hunk 3]
```diff
@@ -398,6 +405,116 @@ func (rh *RouteHandler) GetManifest(response http.ResponseWriter, request *http.
 	WriteData(response, http.StatusOK, mediaType, content)
 }
 
+type ImageIndex struct {
+	ispec.Index
+}
+
+func getReferrers(ctx context.Context, routeHandler *RouteHandler,
```
Comment (peusebiu) [inline pkg/api/routes.go:412]: this is not correct
1. we don't need to sync the subject, regctl will HEAD the subject before getting here, so at this point we can assume we already have the image synced on demand.

2. we need to update sync on demand to copy OCI referrers, like we do for oras. (the case where we have the image synced already but at a time when it didn't have referrers yet on upstream)

4. and also update sync to check each image it syncs if it has OCI referrers.

File: pkg/storage/local/local.go (hunk 1 above)
Comment (rchincha) [reply pkg/storage/local/local.go:1449]:
```
If the registry performs filtering, the response MUST include an annotation (org.opencontainers.references.filtersApplied) denoting that an artifactType filter was applied.
```
Per spec, all it seems to say is that they key be present, value unknown.
https://github.com/opencontainers/wg-reference-types/blob/main/docs/proposals/PROPOSAL_E.md#filtering

File: pkg/extensions/sync/on_demand.go [hunk 4]
```diff
@@ -248,7 +248,7 @@ func syncOneImage(imageChannel chan error, cfg Config, storeController storage.S
 						demandedImageRef, copyErr)
 					time.Sleep(retryOptions.Delay)
 
-					if err = retry.RetryIfNecessary(context.Background(), func() error {
+					if err = retry.RetryIfNecessary(ctx, func() error {
```
[the remark this answers is not in the document]
Comment (rchincha) [reply pkg/extensions/sync/on_demand.go:251]: This is a call we can make, but at least the context being passed in from the http request is desirable.

File: pkg/storage/s3/s3.go (hunk 2 above)
Comment (rchincha) [reply pkg/storage/s3/s3.go:1227]: @peusebiu

File: pkg/api/routes.go (hunk 3 above)
Comment (rchincha) [reply pkg/api/routes.go:412]: @peusebiu pls fix this in a separate PR.

File: pkg/storage/local/local.go (hunk 1 above)
Comment (rchincha) [reply pkg/storage/local/local.go]: That said, may be useful to add this info anyways.

**Review by andaaron — APPROVED** [review]
Looks good for now.

**Pull request merged**